### DISCUSSION TITLE
Does the language of a simulated program affect simulation performance?

### DISCUSSION THREAD

**Discussion by higher-order** [shadow/shadow#3481] [Q&A] [answered]
I have a Python script attached to many Tor clients in a network generated by tornettools. It listens to many Tor control port events per second using the [stem](https://stem.torproject.org/) library, and sends data over HTTP to a custom bandwidth authority (bwauth) per minute using the [requests](https://github.com/psf/requests) library. The custom bwauth is attached with a different Python script that runs the Python Flask server to handle those requests, and does some IO with disk per minute.

I observed that increasing the number of clients that are attached with the client script will significantly increase the time it takes for the simulation to complete. So I'd like to optimize the client script. If I rewrite the exact same functionality using a language faster than Python, such as C, will it take less time for Shadow to run the simulation?

**Comment by stevenengler** [comment]
Here are my thoughts. Other people may have different ideas, since I don’t typically run large simulations.

My guess is that a lot of simulation time is spent handling syscalls, handling events that are queued by syscalls (mainly sleeps, network traffic, eventfd wakeups, etc), or switching between threads or applications. While rewriting the application in a lower level language may help, I’m not sure it would have a large impact. I don't know if anyone has looked closely at the performance of languages like Python in Shadow. What might give a bigger impact is rewriting the application to be simpler, hopefully using fewer syscalls and fewer threads. Running the current application under strace (or with shadow’s strace logging enabled) might give an idea about what syscalls are being used, and maybe you’ll see some that seem excessive and unnecessary for your purposes.

But before doing any of that, it might be useful to get a better idea of where the simulation is spending its time. I would personally start with a simple Tor network. Then add in a client script that just reads the control port events and does nothing else. Then if the runtime performance hasn’t changed much, add a custom bw authority that receives the data and does nothing. And continue adding pieces until you find what significantly slows down the simulation. Hopefully you would be able to test this with a smallish network that only takes less than an hour per run.

**Reply by stevenengler** [reply]
> I have a Python script attached to many Tor clients in a network generated by tornettools. It listens to many Tor control port events per second using the stem library

One cause of the slowdown might be that you're sending a frequent stream of data between tor's control port and your application, which means frequent context switching between tor and the python application. Instead of having your application block waiting for more data and being woken up frequently, it might be better to just repeatedly sleep for maybe 1000 ms and then read all available data on the socket all at once. This means that your application would be woken up once per simulated second rather than many times per second.

**Reply by higher-order** [reply]
My guess is also that the numerous events per second are slowing down the simulation. I'll try optimize that first, and then do the profiling if that doesn't speed things up.

However, with the `stem` library I'm currently using, the way I handle events is by adding listener functions to the stem controller. The code is on stem's tutorial page [here](https://stem.torproject.org/tutorials/tortoise_and_the_hare.html): `controller.add_event_listener(bw_event_handler, EventType.BW)`. So every time stem gets an event from the Tor control port, it will automatically call my listener function. Does this mean that to implement your suggestion of sleeping and then handling events in batch, I will have to either modify stem source code, or not use stem and read from Tor's control port using a socket directly?

**Comment by robgjansen** [comment]
From a Shadow perspective, you have created an unbalanced load scenario with multiple producers (Tor client processes producing Tor control port events) and a single consumer (your python script that processes the control events). Your script is a performance bottleneck in this scenario: even if your script is itself multi-threaded, the "kernel" functionalities that your script uses (ie the syscalls, net io, file io, etc.) will all still have to run on a single shadow "virtual kernel" thread. You could observe the bottleneck visually by watching your CPU load during a simulation (with e.g. `top` or `htop`): you will likely see many periods where a single core runs at high utilization while the other cores are idle or have low utilization. 

A more efficient language would _probably_ not improve things that much, because you are still producing the same amount of work in your code for that single shadow "virtual kernel" thread, and the single shadow thread will not become more efficient at processing the work.

Instead, you want a **distributed design** that fundamentally removes the bottleneck. The goal here is to cut down as much as possible the work required to be done in your single central bwauth script. I suggest engineering things such that each Tor control client "producer" runs a Tor control event processor locally (as another process on the same virtual shadow host) and does most of the heavy lifting in terms of event parsing and processing. Spreading the work across shadow hosts will allow shadow to more effectively parallelize the work. Then those processors could send only a smaller set of information - only the bare minimum things needed - over to the single central bwauth consumer, who could do minimal processing. Balancing the workload like this would be better than changing the programming language of your central script.

**Reply by higher-order** [reply]
> I suggest engineering things such that each Tor control client "producer" runs a Tor control event processor locally (as another process on the same virtual shadow host)

This is already how the simulation is set up currently. The same client script is added to all hosts that is a client (e.g. perfclient10exit, perfclient11exit, ..., markovclient101exit, markovclient102exit, ...) as a process:
```
    - path: /usr/bin/python3
      args: /home/cc/projects/tor-crowd/tor-crowd-sourced-measurement/client-report.py
        /home/cc/projects/tor-crowd/sims/tornets/orig2crowd/7c3dc4fbf944ce03412a81dcd70a4879a7bcbfc0/0.05_10800s/1/conf/client-report.json
      expected_final_state: running
      start_time: 1780
```

That's how each client can listen to the events from its Tor process and send data to the single bwauth. Does the problem of multiple producers and single consumer still exist?

**Reply by robgjansen** [reply]
Hmm, not sure! You still have a single bwauth script that has to process all input from all of the clients, but whether or not that is a bottleneck depends on how much work that script is doing and how balanced that work is over time.

**Comment by higher-order** [comment, marked as answer]
@stevenengler @robgjansen I've got it to work. Thank you both for suggesting to **NOT** rewrite the custom program in a lower-level language. As for the root cause, @stevenengler's intuition was spot on. The stem controller has a reader loop that keeps reading from the socket without delay. My first attempt re-implemented stem's socket reading, but still used the stem controller for authentication. That turned out to not improve anything at all. After I fully removed the stem controller and re-implemented stem's event subscription, authentication (the disabled/empty way), and other functionality using an inner loop that keeps reading data from the socket until all data is read, and an outer loop that adds a few seconds of delay between each read, the simulation (5% network, 3 hours) went from taking 400 hours to only 13 hours. It took just a little more time compared to an original, unmodified simulation. 

Maybe I'll upload my script to a GitHub gist when it's more ready. Thanks again!

**Reply by higher-order** [reply]
Clarification: stem otherwise seems performant in general. It is just trickier in a simulation.

**Reply by stevenengler** [reply]
Glad to hear that you got it working! In the end were you able to modify stem to do this delayed reading, or did you have to write the tor control port code from scratch?

**Reply by higher-order** [reply]
My first attempt added `time.sleep` into stem's reader and event loops, but that prevented any message to be received. So I rewrote just enough of stem's functionality to talk to the Tor control port. I used part of stem's helper functions, and implemented the receiving and parsing myself. It wasn't too hard with stem's implementation in front of me. Here are some snippets:

```python
# Create and connect the socket.
sock = socket.socket(AF_INET, SOCK_STREAM)
#sock.setblocking(False)
sock.connect(("127.0.0.1", 9051))
sock_file = sock.makefile(mode = "rwb")

# Authenticate with the control port.
# Send just "AUTHENTICATE" for disabled control port authentication
# which is what's configured in the sim.
stem.socket.send_message(sock_file, "AUTHENTICATE")
data = stem.socket.recv_message(sock_file)
print(f"Replies for AUTHENTICATE: {data}")

# Subscribe to events.
stem.socket.send_message(sock_file, f"SETEVENTS {event_list}")
data = stem.socket.recv_message(sock_file)
print(f"Replies for SETEVENTS: {data}")

...

# Contiuously read from the socket every second.
# Each read gets all the available data at the socket.
buffer_ = ""
msg_delim = "\r\n"
while True:

    # The inner loop that reads all available data from the socket.
    # If sleep(1) directly in this loop, it will read a fixed (small)
    # amount of data every second, instead of all available data.
    while True:

        # Use a timeout of 0 to poll without blocking.
        rs, _, _ = select([sock], [], [], 0)

        if sock in rs:
            raw_data = sock.recv(10240)
            data = stem.util.str_tools._to_unicode(raw_data)
            buffer_ += data
        # If `select` shows nothing to be read at the socket,
        # break out of the reading loop.
        else:
            break

    # If at least 1 message is received, process them.
    if msg_delim in buffer_:

        msgs = buffer_.split(msg_delim)
        # Reset the buffer, saving the trailing partial message if there is any.
        buffer_ = msgs[-1]
        # Leave out the trailing empty or partial message.
        msgs = msgs[:-1]

        for msg in msgs:
            # Convert each message into a `stem.response.ControlMessage`.
            status_code, divider, content = msg[:3], msg[3:4], msg[4:]
            ctrl_msg = stem.response.ControlMessage(
                [(status_code, divider, content)],
                msg,
                None
            )
            stem.response.convert("EVENT", ctrl_msg)

            # TODO: Is it necessary to ensure that a ControlMessage
            #       in our case is always an event, even though we
            #       only subscribed to the CIRC_BW event?
            event = ctrl_msg

            # If it's a CIRC event:
            if isinstance(event, stem.response.events.CircuitEvent):

            ...

    ...

    # Sleep in the outer loop.
    sleep(1)
```
I can publish the full version when it's more ready.
